src/systems/systems/src/shared/pid.rs

```diff
@@ -0,0 +1,166 @@
+use std::time::Duration;
+
+// Pid controller implementation
+// Implementation in a recursive form
+// u(k+1) = u(k) + (e(k)-e(k-1))*Kp + e(k) * Ki * dt + (e(k)- 2*e(k-1) + e(k-2)) * Kd / dt
+// For variable dt add dt duration as argument, if fixed dt or scheduled synchronously with controlled system input None as dt
+pub struct Pid {
+    kp: f64,
+    ki: f64,
+    kd: f64,
+
+    min_output: f64,
+    max_output: f64,
+
+    setpoint: f64,
+
+    error_k_1: Option<f64>,
+    error_k_2: Option<f64>,
+    output: f64,
+}
+
+impl Pid {
+    pub fn new(kp: f64, ki: f64, kd: f64, min_output: f64, max_output: f64, setpoint: f64) -> Self {
+        Self {
+            kp,
+            ki,
+            kd,
+            min_output,
+            max_output,
+            setpoint,
+            error_k_1: None,
+            error_k_2: None,
+            output: 0.,
```

Comment:
This one is pretty much on purpose, as for None error we just skip d term evaluation whereas for first iteration the output really should be 0.  Yet I see how it can get confusing...... Will have a second look!

Reply:
Ok, I guess I can see how you'd want to have `output` be `0` if `next_control_output` has not been called yet. 👍